pkg/action/lazyclient.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package action
+
+import (
+	"context"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type lazyClient struct {
+	// client caches an initialized kubernetes client
+	client     kubernetes.Interface
+	initClient sync.Once
+
+	// clientFn loads a kubernetes client
+	clientFn func() (*kubernetes.Clientset, error)
+
+	namespace string
+}
+
+func (s *lazyClient) init() error {
+	var err error
+	s.initClient.Do(func() {
+		if s.client == nil {
+			s.client, err = s.clientFn()
```

Comment:
lazyClient should probably cache any client initialization error as well… if init() errors on first call, subsequent calls will return nil, but s.client will still be nil, leading to a NPE